libminifi/src/controllers/SSLContextService.cpp

```diff
@@ -128,16 +149,239 @@ bool SSLContextService::configure_ssl_context(SSL_CTX *ctx) {
+                                        CERT_STORE_OPEN_EXISTING_FLAG | CERT_STORE_READONLY_FLAG | store_location.getBitfieldValue(),
+                                        client_cert_store_.data());
+  if (!hCertStore) {
+    logger_->log_error("Could not open system certificate store %s/%s (client certificates)", cert_store_location_, client_cert_store_);
+    return false;
+  }
+  const auto store_close = gsl::finally([hCertStore](){ CertCloseStore(hCertStore, 0); });
+
+  logger_->log_debug("Looking for client certificate in sytem store %s/%s", cert_store_location_, client_cert_store_);
+
+  PCCERT_CONTEXT pCertContext = nullptr;
+  while (pCertContext = CertEnumCertificatesInStore(hCertStore, pCertContext)) {
+    bool certificateIsAcceptableAndWasSuccessfullyAddedToSSLContext = useClientCertificate(ctx, pCertContext);
+    if (certificateIsAcceptableAndWasSuccessfullyAddedToSSLContext) {
+      CertFreeCertificateContext(pCertContext);
+      return true;
+    }
+  }
+
+  logger_->log_error("Could not find any suitable client certificate in sytem store %s/%s", cert_store_location_, client_cert_store_);
+  return false;
+#else
+  logger_->log_error("Getting client certificate from the system store is only supported on Windows");
+  return false;
+#endif  // WIN32
+}
+
+#ifdef WIN32
+bool SSLContextService::useClientCertificate(SSL_CTX* ctx, PCCERT_CONTEXT certificate) const {
+  utils::tls::X509_unique_ptr x509_cert = utils::tls::convertWindowsCertificate(certificate);
+  if (!x509_cert) {
+    logger_->log_error("Failed to convert system store client certificate to X.509 format");
+    return false;
+  }
+
+  utils::tls::EVP_PKEY_unique_ptr private_key = utils::tls::extractPrivateKey(certificate);
+  if (!private_key) {
+    logger_->log_debug("Skipping client certificate %s because it has no exportable private key", x509_cert->name);
+    return false;
+  }
+
+  if (!client_cert_cn_.empty()) {
+    utils::tls::DistinguishedName dn = utils::tls::DistinguishedName::fromSlashSeparated(x509_cert->name);
+    utils::optional<std::string> cn = dn.getCN();
+    if (!cn || *cn != client_cert_cn_) {
+      logger_->log_debug("Skipping client certificate %s because it doesn't match CN=%s", x509_cert->name, client_cert_cn_);
+      return false;
+    }
+  }
+
+  utils::tls::EXTENDED_KEY_USAGE_unique_ptr key_usage{static_cast<EXTENDED_KEY_USAGE*>(X509_get_ext_d2i(x509_cert.get(), NID_ext_key_usage, nullptr, nullptr))};
+  if (!key_usage) {
+    logger_->log_error("Skipping client certificate %s because it has no extended key usage", x509_cert->name);
+    return false;
+  }
+
+  if (!(client_cert_key_usage_.isSubsetOf(utils::tls::ExtendedKeyUsage{*key_usage}))) {
+    logger_->log_debug("Skipping client certificate %s because its extended key usage set does not contain all usages specified in %s",
+                       x509_cert->name, Configuration::nifi_security_windows_client_cert_key_usage);
+    return false;
+  }
```

Comment:
It would, if (1) I had known about it, and (2) this function was available in libressl.  This is a relatively new addition in openssl: https://github.com/openssl/openssl/commit/063f1f0c693a10aab6a7227df15d4120ed824856 and is not in libressl, yet: https://github.com/libressl-portable/openbsd/blob/master/src/lib/libcrypto/x509/x509v3.h#L789

I think we should ditch libressl and switch to openssl when 3.0.0 is released, as that will (finally) be under an Apache license.

Reply:
Oh, that explains it, thanks. The switching sounds like a mid to longterm objective, so I'm marking this resolved now.